### PR TITLE
Master mail handle exceptions activity mpa

### PR DESCRIPTION
**Pull request by mitali-willdooit**
Description of the issue/feature this PR addresses:
<b> Task:</b> https://www.odoo.com/web?#id=1889379&action=333&active_id=131&model=project.task&view_type=form&menu_id=4720
<b> Pad:</b> https://pad.odoo.com/p/r.493858ed67f383dd21b6334fde8c6050
Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/mail/models/mail_activity.py [hunk 1]
```diff
@@ -601,6 +601,16 @@ class MailActivityMixin(models.AbstractModel):
...
         search='_search_activity_summary',
         groups="base.group_user",)
+    exception_activity_ids = fields.One2many(
+        'mail.activity', 'res_id', 'Exception Activities',
+        auto_join=True,
+        groups="base.group_user",
+        domain=lambda self: [('res_model', '=', self._name), ('activity_type_id.decoration_type', 'in', ['warning', 'danger'])])
+    exception_activity_type = fields.Char(string=' ', compute='_compute_exception_activity_type')
+
+    def _compute_exception_activity_type(self):
```
Comment (Xavier-Do) [inline addons/mail/models/mail_activity.py]: After a quick look, I would remove the exception_activity_ids field, base the exception_activity_type on activity_ids.
It could be a stored computed field BUT invalidation could be expensive since the dependency is very specific. 

I think that the most effective solution would be to do that manually, just add a exception_activity_type field on mixin and maintains his value on activity write/create 
pseudo code: 
write/create/unlink for activity:
if self.env[model].browse('res_id') .is_mail_activity and  activity_type in values or active in values:
      self.env[model].browse('res_id') ._compute_exception_activity_type

_compute_exception_activity_type should be based on activity_type with filters instead. 

Just a quick draft, feel free to adapt names, logic.

File: addons/mail/models/mail_activity.py [hunk 2]
```diff
@@ -631,6 +643,13 @@ def _search_activity_user_id(self, operator, operand):
     def _search_activity_type_id(self, operator, operand):
         return [('activity_ids.activity_type_id', operator, operand)]
 
+    @api.multi
+    def set_exception_activity_type(self):
```
Comment (Xavier-Do) [inline addons/mail/models/mail_activity.py]: could be private no?

File: addons/mail/models/mail_activity.py [hunk 3]
```diff
@@ -328,6 +328,8 @@ def create(self, values):
             self.env['bus.bus'].sendone(
                 (self._cr.dbname, 'res.partner', activity.user_id.partner_id.id),
                 {'type': 'activity_updated', 'activity_created': True})
+        if activity.res_model and activity.res_id:
+            self.env[activity.res_model].browse(activity.res_id).set_exception_activity_type()
```
Comment (Xavier-Do) [inline addons/mail/models/mail_activity.py]: 1. We could avoid to set_exception_activity_type if activity_type_id.decoration_type not in ['warning', 'danger'] in create.
-> create only, write and unlink need to set_exception_activity_type too if we remove a warning/danger activity type.

File: addons/mail/models/mail_activity.py [hunk 4]
```diff
@@ -631,6 +643,13 @@ def _search_activity_user_id(self, operator, operand):
     def _search_activity_type_id(self, operator, operand):
         return [('activity_ids.activity_type_id', operator, operand)]
 
+    @api.multi
+    def set_exception_activity_type(self):
+        self.ensure_one()
+        self.exception_activity_type = False
```
Comment (Xavier-Do) [inline addons/mail/models/mail_activity.py]: - This is a first write that can be useless
- writing on exception_activity_type in a loop. 
- will keep the last activity_type_id, why the last? We could break on first, or keep the more important one
- "for record" can be "for activity" to improve readability. 
Improvement suggestion:
``` python3
decoration_type = None
icon = None
for activity in self.activity_ids.filte....
    if not decoration and activity.activity_type_id.decoration_type == 'warning'
        decoration_type = 'warning'
        icon = record.activity_type_id.icon
    elif activity.activity_type_id.decoration_type == 'danger'
        decoration_type = 'warning'
        icon = record.activity_type_id.icon
        break
self.exception_activity_type = "pull-right text-%s fa %s" % (decoration_type, icon)
```
What do you think?

File: addons/mail/models/mail_activity.py [hunk 5]
```diff
@@ -631,6 +643,13 @@ def _search_activity_user_id(self, operator, operand):
...
+        for record in self.activity_ids.filtered(lambda x: x.activity_type_id.decoration_type in ['warning', 'danger']):
+            self.exception_activity_type = "pull-right text-%s fa %s" % (record.activity_type_id.decoration_type, record.activity_type_id.icon)
```
[the remark this answers is not in the document]
Comment (mge-odoo) [reply addons/mail/models/mail_activity.py]: Oh, I check this morning and this is the case so we should probably go for the second solution.

File: addons/mail/models/mail_activity.py [hunk 6]
```diff
@@ -601,6 +612,7 @@ class MailActivityMixin(models.AbstractModel):
         related='activity_ids.summary', readonly=False,
         search='_search_activity_summary',
         groups="base.group_user",)
+    exception_activity_type = fields.Char(string=' ')
```
[the remark this answers is not in the document]
Comment (mge-odoo) [reply addons/mail/models/mail_activity.py]: I would also add a comment section explaining why we didn't use the ORM feature (see previous xdo review about performance).

File: addons/mail/static/src/js/activity.js [hunk 7]
```diff
@@ -681,6 +682,7 @@ var KanbanActivity = BasicActivity.extend({
      * @private
      */
     _render: function () {
+        this.renderElement()
```
Comment (mge-odoo) [inline addons/mail/static/src/js/activity.js]: why this change?

File: addons/mail/static/src/js/activity.js [hunk 8]
```diff
@@ -740,9 +742,23 @@ var KanbanActivity = BasicActivity.extend({
     },
 });
 
+// -----------------------------------------------------------------------------
+// Activity Exception Widget to display Exception icon ('activity_exception' widget)
+// -----------------------------------------------------------------------------
+var ActivityException = basic_fields.FieldChar.extend({
+
+    _renderReadonly: function () {
```
Comment (mge-odoo) [inline addons/mail/static/src/js/activity.js]: so in edit the field is just an input? Don't you always want to render it in readonly?

File: addons/mail/models/mail_activity.py [hunk 9]
```diff
@@ -631,6 +643,20 @@ def _search_activity_user_id(self, operator, operand):
...
+        icon = None
+        for activity in self.activity_ids.filtered(lambda x: x.activity_type_id.decoration_type in ['warning', 'danger']):
+            if not decoration_type and activity.activity_type_id.decoration_type == 'warning':
+                decoration_type = 'warning'
+                icon = activity.activity_type_id.icon
+            elif activity.activity_type_id.decoration_type == 'danger':
+                decoration_type = 'danger'
+                icon = activity.activity_type_id.icon
+                break
+        self.exception_activity_type = "pull-right text-%s fa %s" % (decoration_type, icon)
```
Comment (mge-odoo) [inline addons/mail/models/mail_activity.py]: I have to admit that I don't like this server-side formatting at all. This should be the role of the client (deciding how to display and which classes apply).

Possible solutions:
 * still keep one char field `warning,fa-tasks`, and the widget splits the value. This is still not perfect because one field has to carry multiple information, and you may have inconsistency
 * two fields 1) field containing `warning` (might be a selection instead) and 2) second field containg`fa-tasks` ; the JS widget would be on one of them (I'd go for the first field, because this is the one making sense, the second one is more like a complement of information) and the JS widget could use `fieldDependencies` property (see https://github.com/odoo/odoo/blob/master/addons/web/static/src/js/fields/abstract_field.js#L50 so you don't have to add the field in invisible in every view) to fetch the second field ; this proposition seems to be the best according to me but performance-wise, one might check if the `fieldDependencies` are correctly read in batch.

Comment (mge-odoo) [reply addons/mail/models/mail_activity.py]: Oh, I check this morning and this is the case so we should probably go for the second solution.

File: addons/mail/models/mail_activity.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (mpa-odoo) [reply addons/mail/models/mail_activity.py]: @mge-odoo Can we go for the second solution?

[the remark this answers is not in the document]
Comment (mge-odoo) [reply addons/mail/models/mail_activity.py]: Yes you can, thank you.

File: addons/mail/static/src/js/activity.js (hunk 7 above)
Comment (mpa-odoo) [reply addons/mail/static/src/js/activity.js]: @mge-odoo when kanban view is not group By with any field and we are scheduling the new activity then whole kanban view will not be rendered so icon is not going to update instantly so we need to re-render the element by using renderElement() or we may change the icon explicitly by selector

File: addons/test_mail/tests/test_performance.py [hunk 10]
```diff
@@ -182,7 +182,7 @@ def test_adv_activity_full(self):
             #voip module read activity_type during create leading to one less query in enterprise on action_feedback
             category = activity.activity_type_id.category
 
-        with self.assertQueryCount(__system__=24, emp=41):  # com runbot: 24 - 41 // test_mail only: 24 - 39
+        with self.assertQueryCount(__system__=30, emp=47):  # com runbot: 24 - 41 // test_mail only: 24 - 39
```
Comment (Xavier-Do) [inline addons/test_mail/tests/test_performance.py]: ok we will need to improve that

File: addons/mail/models/mail_activity.py [hunk 11]
```diff
@@ -355,16 +357,25 @@ def write(self, values):
...
                             (self._cr.dbname, 'res.partner', partner.id),
                             {'type': 'activity_updated', 'activity_deleted': True})
+        if values.get('activity_type_id'):
+            for activity in self.filtered(lambda x: x.res_id and x.res_model):
+                self.env[activity.res_model].browse(activity.res_id)._set_activity_exception_type()
         return res
 
     @api.multi
     def unlink(self):
+        record_values = []
```
Comment (Xavier-Do) [inline addons/mail/models/mail_activity.py]: record_values could be renamed + self.env[record[0]].browse(record[1]) is no very elegant.
you could immediately add the browse in the list

File: addons/mail/models/mail_activity.py [hunk 12]
```diff
@@ -328,6 +328,8 @@ def create(self, values):
             self.env['bus.bus'].sendone(
                 (self._cr.dbname, 'res.partner', activity.user_id.partner_id.id),
                 {'type': 'activity_updated', 'activity_created': True})
+        if activity.res_model and activity.res_id and activity.activity_type_id.decoration_type in ['warning', 'danger']:
+            self.env[activity.res_model].browse(activity.res_id)._set_activity_exception_type()
```
Comment (Xavier-Do) [inline addons/mail/models/mail_activity.py]: We actually have no certitude that self.env[activity.res_model].browse(activity.res_id) is a mail_activity_mixin i think.

File: addons/mail/models/mail_activity.py [hunk 13]
```diff
@@ -355,16 +357,25 @@ def write(self, values):
...
                             (self._cr.dbname, 'res.partner', partner.id),
                             {'type': 'activity_updated', 'activity_deleted': True})
+        if values.get('activity_type_id'):
+            for activity in self.filtered(lambda x: x.res_id and x.res_model):
+                self.env[activity.res_model].browse(activity.res_id)._set_activity_exception_type()
         return res
 
     @api.multi
     def unlink(self):
+        activitiy_records = []
```
Comment (Xavier-Do) [inline addons/mail/models/mail_activity.py]: should be a set, we only need to _set_activity_exception_type once per record.

File: addons/mail/models/mail_activity.py [hunk 14]
```diff
@@ -355,16 +357,25 @@ def write(self, values):
                         self.env['bus.bus'].sendone(
                             (self._cr.dbname, 'res.partner', partner.id),
                             {'type': 'activity_updated', 'activity_deleted': True})
+        if values.get('activity_type_id'):
```
Comment (Xavier-Do) [inline addons/mail/models/mail_activity.py]: Not sure we are writing res_model or res_id on activities, but to be complete, we would need call _set_activity_exception_type on record before write, in case the activity is switch to another one. Not sure it is a use case we need to support, not a big deal if we don't do that.

File: addons/mail/models/mail_activity.py [hunk 15]
```diff
@@ -355,16 +357,25 @@ def write(self, values):
...
     @api.multi
     def unlink(self):
+        activitiy_records = []
         for activity in self:
             if activity.date_deadline <= fields.Date.today():
                 self.env['bus.bus'].sendone(
                     (self._cr.dbname, 'res.partner', activity.user_id.partner_id.id),
                     {'type': 'activity_updated', 'activity_deleted': True})
-        return super(MailActivity, self).unlink()
+            if activity.res_model and activity.res_id:
```
Comment (Xavier-Do) [inline addons/mail/models/mail_activity.py]: and activity.activity_type_id.decoration_type in ['warning', 'danger']:

unlink is the reason why we have broken query count, i'd like to try to improve this a little bit.
If we remove an activity of another type, the decoration_type won't change. If decoration_type was warning or danger, it was because of another activity. If decoration_type was none, we don't care. Am I right?

Can you try this and check the new query count?

File: addons/mail/models/mail_activity.py (hunk 14 above)
Comment (mba-odoo) [reply addons/mail/models/mail_activity.py]: @Xavier-Do , If we use before the write then activity_type_id will not be the right(it will be older one). and we used `activity_type_id` in `_set_activity_exception_type` Thanks

Comment (Xavier-Do) [reply addons/mail/models/mail_activity.py]: I mean that if I write a res_id on an activity, the previously linked record will loose an activity. To be complete, we would need to do something like.
record_to_update= set()
if res_model  in values or res_id in values:
    for activity in self....
         record_to_update.append...
self.write....
if res_model  in values or res_id in values or activity_type_id in values:
    for activity in self....
         record_to_update.append...
for rec in record_to_update:
    rec.....

But since we are not switching activities from one record to another, it's maybe not the most necessary to do so.

File: addons/mail/static/src/js/activity.js [hunk 16]
```diff
@@ -740,9 +752,49 @@ var KanbanActivity = BasicActivity.extend({
     },
 });
 
+// -----------------------------------------------------------------------------
+// Activity Exception Widget to display Exception icon ('activity_exception' widget)
+// -----------------------------------------------------------------------------
+
+var ActivityException = AbstractField.extend({
```
Comment (mge-odoo) [inline addons/mail/static/src/js/activity.js]: Can you add some QUnit tests for this new widget?

File: addons/mail/static/src/js/activity.js [hunk 17]
```diff
@@ -682,6 +686,14 @@ var KanbanActivity = BasicActivity.extend({
      */
     _render: function () {
         var $span = this.$('.o_activity_btn > span');
+        // when no goupby on kanban, it will not call renderElement so update activity icon manually.
+        if (this.recordData.decoration_type) {
+            $span.removeClass('fa-clock-o');
```
Comment (mge-odoo) [inline addons/mail/static/src/js/activity.js]: I would like to see tests (or test modifications) that express the new behaviors (see chatter_tests.js).

File: addons/mail/models/mail_activity.py [hunk 18]
```diff
@@ -603,6 +614,11 @@ class MailActivityMixin(models.AbstractModel):
         related='activity_ids.summary', readonly=False,
         search='_search_activity_summary',
         groups="base.group_user",)
+    decoration_type = fields.Selection([
```
Comment (mge-odoo) [inline addons/mail/models/mail_activity.py]: As we are in the activity mixin, these two fields must be prefixed with `activity_` (otherwise they could collide with existing fields).

File: addons/mail/models/mail_activity.py [hunk 19]
```diff
@@ -603,6 +614,11 @@ class MailActivityMixin(models.AbstractModel):
...
+        ('warning', 'Alert'),
+        ('danger', 'Error')],
+        help="Change the background color of the related activities of this type.")
```
Comment (mge-odoo) [inline addons/mail/models/mail_activity.py]: The `help` attribute doesn't not express what the field is meant for.

File: addons/mail/models/mail_activity.py [hunk 20]
```diff
@@ -603,6 +614,11 @@ class MailActivityMixin(models.AbstractModel):
...
+    icon = fields.Char('Icon', help="This icon indicates that exception activity planned on an object.")
```
Comment (mge-odoo) [inline addons/mail/models/mail_activity.py]: same about `help`

File: addons/mail/models/mail_activity.py (hunk 18 above)
Comment (mge-odoo) [reply addons/mail/models/mail_activity.py]: I would also use more explicit names, like `activity_exception_decoration` and `activity_exception_icon`.

Comment (mge-odoo) [reply addons/mail/models/mail_activity.py]: And again, I would like to see a comment explaining why we bypass ORM features for performance reasons on these fields.

File: addons/mail/static/src/js/activity.js [hunk 21]
```diff
@@ -682,6 +686,14 @@ var KanbanActivity = BasicActivity.extend({
      */
     _render: function () {
         var $span = this.$('.o_activity_btn > span');
+        // when no goupby on kanban, it will not call renderElement so update activity icon manually.
```
[the remark this answers is not in the document]
Comment (mpa-odoo) [reply addons/mail/static/src/js/activity.js]: @mge-odoo It doesn't work with the kanban so we keep as it is. Thanks.

File: addons/mail/static/src/js/activity.js (hunk 16 above)
Comment (mpa-odoo) [reply addons/mail/static/src/js/activity.js]: @mge-odoo  Done.

File: addons/mail/static/src/js/activity.js (hunk 17 above)
Comment (mpa-odoo) [reply addons/mail/static/src/js/activity.js]: @mge-odoo Done.

File: addons/mail/models/mail_activity.py (hunk 18 above)
Comment (mpa-odoo) [reply addons/mail/models/mail_activity.py]: @mge-odoo Done

File: addons/mail/static/src/xml/web_kanban_activity.xml [hunk 22]
```diff
@@ -4,7 +4,7 @@
 <t t-name="mail.KanbanActivity">
     <div class="o_kanban_inline_block dropdown o_kanban_selection o_mail_activity">
         <a class="dropdown-toggle o-no-caret o_activity_btn" data-toggle="dropdown" role="button">
-            <span class="fa fa-clock-o fa-lg fa-fw" t-att-title="widget.selection[widget.activityState]" role="img" t-att-aria-label="widget.selection[widget.activity_state]"/>
+            <span t-attf-class="fa-lg fa-fw fa #{widget.recordData.activity_exception_decoration ? 'pull-right text-'+ widget.recordData.activity_exception_decoration +' '+ widget.recordData.activity_exception_icon : 'fa-clock-o'}" t-att-title="widget.selection[widget.activityState]" role="img" t-att-aria-label="widget.selection[widget.activity_state]"/>
```
Comment (mge-odoo) [inline addons/mail/static/src/xml/web_kanban_activity.xml]: why do you add class here if you already do it in JS (see https://github.com/odoo/odoo/pull/31839/files#diff-75c502c511f53d668928d69e9341d308R692 )

File: addons/mail/static/tests/chatter_tests.js [hunk 23]
```diff
@@ -650,13 +675,15 @@ QUnit.test('kanban activity widget with an activity', async function (assert) {
         "activity widget should have been rendered correctly");
     assert.strictEqual(rpcCount, 1, '1 RPC (search_read) should have been done');
 
+    // Check exception activity icon fa-warning
+    assert.hasClass($record.find('.o_activity_btn span'), "fa-warning", "should display correct exception activity icon");
     // click on the activity button
     await testUtils.dom.click($record.find('.o_activity_btn'));
     assert.strictEqual(rpcCount, 2, 'a read should have been done to fetch the activity details');
-    assert.strictEqual($record.find('.o_activity_title').length, 1, "should have an activity scheduled");
+    assert.strictEqual($record.find('.o_activity_title').length, 2, "should have an activity scheduled");
```
Comment (mge-odoo) [inline addons/mail/static/tests/chatter_tests.js]:
```suggestion
    assert.strictEqual($record.find('.o_activity_title').length, 2, "should have two activities scheduled");
```

File: addons/mail/static/tests/chatter_tests.js [hunk 24]
```diff
@@ -666,11 +693,19 @@ QUnit.test('kanban activity widget with an activity', async function (assert) {
     await testUtils.dom.click($record.find('.o_activity_btn'));
     assert.strictEqual(rpcCount, 3, 'should have reloaded the activities');
 
+    // mark exception activity as done
+    await testUtils.dom.click($record.find('.o_mark_as_done:eq(1)'));
+    await testUtils.dom.click($record.find('.o_activity_popover_done'));
+    $record = kanban.$('.o_kanban_record').first(); // the record widget has been reset
+
+    // check activity icon replace with fa-clock
```
Comment (mge-odoo) [inline addons/mail/static/tests/chatter_tests.js]:
```suggestion
    // check activity icon replaced with fa-clock
```

File: addons/mail/static/tests/chatter_tests.js [hunk 25]
```diff
@@ -738,6 +773,60 @@ QUnit.test('kanban activity widget popover test', async function (assert) {
...
+        View: ListView,
+        model: 'partner',
+        data: this.data,
+        arch: '<tree>' +
+                '<field name="foo"/>'+
+                '<field name="activity_exception_decoration" widget="activity_exception"/> ' +
+            '</tree>',
+    });
+
+    assert.strictEqual(list.$('.o_data_row').length, 2, "Should have two record");
```
Comment (mge-odoo) [inline addons/mail/static/tests/chatter_tests.js]:
```suggestion
    assert.containsN(list, '.o_data_row', 2, "should have two records");
```

File: addons/sale/views/sale_views.xml [hunk 26]
```diff
@@ -155,6 +155,7 @@
                     <field name="state"/>
                     <field name="currency_id"/>
                     <field name="activity_state"/>
+                    <field name="activity_ids"/>
```
Comment (mge-odoo) [inline addons/sale/views/sale_views.xml]: no need to declare it before

File: addons/stock/views/stock_picking_views.xml [hunk 27]
```diff
@@ -151,6 +151,7 @@
                     <field name="state"/>
                     <field name="date"/>
                     <field name="activity_state"/>
+                    <field name="activity_ids"/>
```
Comment (mge-odoo) [inline addons/stock/views/stock_picking_views.xml]: no need to declare it before

File: addons/mrp/views/mrp_production_views.xml [hunk 28]
```diff
@@ -196,6 +197,7 @@
                     <field name="date_planned_start"/>
                     <field name="state"/>
                     <field name="activity_state"/>
+                    <field name="activity_ids"/>
```
Comment (mge-odoo) [inline addons/mrp/views/mrp_production_views.xml]: no need to declare it before ; it is only useful when your kanban-box template uses `t-esc="record.<fieldName>.value"` somewhere

File: addons/purchase/views/purchase_views.xml [hunk 29]
```diff
@@ -343,6 +345,7 @@
                     <field name="date_order"/>
                     <field name="currency_id"/>
                     <field name="activity_state"/>
+                    <field name="activity_ids"/>
```
Comment (mge-odoo) [inline addons/purchase/views/purchase_views.xml]: no need to declare it before

File: addons/mail/static/src/js/activity.js [hunk 30]
```diff
@@ -682,6 +686,14 @@ var KanbanActivity = BasicActivity.extend({
...
+        if (this.recordData.activity_exception_decoration) {
+            $span.removeClass('fa-clock-o');
+            $span.addClass('text-' + this.recordData.activity_exception_decoration + ' ' + this.recordData.activity_exception_icon);
+        } else {
+            $span.removeClass();
+            $span.addClass('fa-lg fa-fw fa fa-clock-o');
```
Comment (mge-odoo) [inline addons/mail/static/src/js/activity.js]: Mmmmh I don't like this ; if you change a class in the template, you also have to repeat the change here, which is not ideal.

Instead, I would use two different `<span>` in the template (you can use a special class to identify them) and toggle them according to `activity_exception_decoration`.

**Comment by mge-odoo** [comment]
Also note that some views are in conflict ; you might need to rebase your branches on master.

File: addons/mail/static/src/js/activity.js (hunk 30 above)
Comment (mpa-odoo) [reply addons/mail/static/src/js/activity.js]: @mge-odoo Here in each icon is different base on an activity type and we need to update icon accordingly so we haven't any particular class to remove.

File: addons/mail/models/mail_activity.py [hunk 31]
```diff
@@ -603,6 +614,13 @@ class MailActivityMixin(models.AbstractModel):
         related='activity_ids.summary', readonly=False,
         search='_search_activity_summary',
         groups="base.group_user",)
+    # Instead of fetching decoration_type and icon each time from activity type
```
[the remark this answers is not in the document]
Comment (mpa-odoo) [reply addons/mail/models/mail_activity.py]: @mge-odoo Done

File: addons/mail/static/src/xml/web_kanban_activity.xml (hunk 22 above)
Comment (mpa-odoo) [reply addons/mail/static/src/xml/web_kanban_activity.xml]: @mge-odoo Done

File: addons/hr/models/hr_employee.py [hunk 32]
```diff
@@ -126,6 +126,8 @@ def _default_image(self):
     ], string="Departure Reason", groups="hr.group_hr_user", copy=False, tracking=True)
     departure_description = fields.Text(string="Additional Information", groups="hr.group_hr_user", copy=False, tracking=True)
     message_main_attachment_id = fields.Many2one(groups="hr.group_hr_user")
+    activity_exception_decoration = fields.Selection(groups="hr.group_hr_user")
```
[the remark this answers is not in the document]
Comment (mpa-odoo) [reply addons/hr/models/hr_employee.py]: @mge-odoo Only these two fields are the only storable fields of mail.activity.mixin, other fields are not storable fields so no need to add here.

File: addons/mail/static/src/js/activity.js (hunk 30 above)
Comment (mge-odoo) [reply addons/mail/static/src/js/activity.js]: @mpa-odoo the special class I mentioned was just meant to be able to identify and differentiate the two `<span>` in the template, like I mentioned (like `<span class='o_activity_with_exception ...'/>` and `<span class='o_activity_without_exception fa fa-clock-o fa-lg fa-fw ...'/>`) ; do you understand why I want to `<span>`?

Comment (mpa-odoo) [reply addons/mail/static/src/js/activity.js]: @mge-odoo  yeah , got it. Thank you

**Comment by mge-odoo** [comment]
Alright, the branches seem ready to be merged.

@Xavier-Do do you confirm the stored fields on the mixin?
@mpa-odoo would you mind rebase your branches to resolve the conflicts please?

**Comment by Xavier-Do** [comment]
After analyse with odony, we think that the field may be computed. The activity ratio is actually quite low. The trade-off hardrive/cpu time here seems to show that a compute field could have good enough performances and avoid to increase database size too mush. 
Sorry for the back and forward.

**Comment by Xavier-Do** [comment]
After discussion with @mge-odoo, we will take care of the last details in order to merge it asap. Thanks for your work!

**Comment by Xavier-Do** [comment]
robodoo r+

**Comment by Xavier-Do** [comment]
@robodoo r+ rebase-ff
